protobuf.bzl

```diff
@@ -526,6 +555,137 @@ def py_proto_library(
         **kargs
     )
 
+def _source_proto_library(
```

Comment:
If we pass outs as a param here so that we can pass in the language-specific outs, can we use this for all languages and combine it with proto_gen so we don't need two separate ones?

Reply:
What do you mean by "two separate ones"?

This macro is specifically for generated proto code we don't know how to compile, so it just bundles them up in a filegroup.  This should probably be the first one to go, replaced with language-specific macros that actually do some kind of compilation.

Reply:
Grouping all the protobuf.bzl design questions here because they kind of come together. 

My design idea here is:
- Make `_source_proto_library` accept a list of outs in addition to the language
- All language-specific rules call that
- The ones we want to compile (python and objc) compile the filegroup that's produced
- Maybe allows a combination with `_source_proto_library` and `proto_gen`

I'm hoping this could solve the objc problem because we can pass in the headers/sources separately. 

It seems like there should be a cleaner solution here, but if this idea also doesn't work it's fine to keep as is especially since this should be temporary.

Reply:
hmm that does sound nicer.  Lemme try it out and see if I can get it to work :)

Reply:
ah trying this out I remembered the issue here (I think I did try something similar).  _proto_gen's implementation is a little awkward where it runs protoc over each src file individually (and sometimes over each src/output pair).  So it needs to know the mapping from src->outs, and a static list of outs won't work.

That being said, now that everything's green I can try to just run a single protoc command and see if anything breaks :)

Reply:
Yea I don't think this will work in general.  The problem is if someone passes multiple protos in different directories to these rules.  We need to be able to call protoc separately for each directory, so we need to know which outputs correspond to which src.  We could ban this case, but that would make the macros functionally worse for some internal cleanup

Reply:
That makes sense, thanks for giving it a try!